### ISSUE TITLE
Complexity issue on 5.4+ using kubeProxyReplacement=disabled + IPSec

### ISSUE THREAD

**Issue by msibrichins** [cilium/cilium#14784]
## Bug report

**General Information**

- Cilium version: 1.8.6 and 1.9.3
- Kernel version: 5.4.0
- Orchestration system version in use: kubectl and kubeadm v1.16.15, Ubuntu 20.04, 
- Generate and upload a system zip: 
[cilium-sysdump-20210128-160233.zip](https://github.com/cilium/cilium/files/5890463/cilium-sysdump-20210128-160233.zip)

**How to reproduce the issue**

1. Setup a VM on subnet 10.1.0.0/19
2. Install kubectl, kubelet, kube-apiserver, etc. then run `kubeadm init --config kubeadm-config.yaml`
[kubeadm-config.yaml.gz](https://github.com/cilium/cilium/files/5890484/kubeadm-config.yaml.gz)

3. Wait for kubernetes to come up.  Watch the pods in the kube-system namespace.  You should see kube-apiserver, kube-controller-manager, kube-scheduler, and kube-proxy up and running.  You should see two coredns pods in a pending state.  `kubectl get nodes` should show a single node in a not ready state.
2. Helm install cilium.  In this case using kubernetes as the ipam since the pod network and subnet are specified in the kubeadm config file.
```
helm install cilium cilium/cilium --version 1.8.6 \
  --namespace kube-system \
  --set global.kubeProxyReplacement=disabled \
  --set config.ipam="kubernetes" \
  --set global.encryption.enabled=true \
  --set global.encryption.nodeEncryption=false 
```
3. Wait for the cilium agent to deploy.  When successful, the core-dns pods become scheduled and ready after the cilium agent comes up.  In the case of this bug the core-dns pods change status from Pending to ContainerCreating (observed with `kubectl get pods -n kube-system`)
4. `kubectl describe pod -n kube-system coredns-<id>` indicates the CNI failed to setup pod. 
[describe-coredns-pod.txt](https://github.com/cilium/cilium/files/5890430/describe-coredns-pod.txt)

Logs from the cilium agent:
[cilium-agent.log.gz](https://github.com/cilium/cilium/files/5890493/cilium-agent.log.gz)

The azure.json file is not specified but it's a vanilla azure k8s json file.  Let me know if you think it's needed and I'll attempt to redact then upload it appropriately.

**Comment by msibrichins** [comment]
Thanks @pchaigno for the analysis.  I'm trying to understand what a "complexity issue" is.  I read through your post at https://pchaigno.github.io/ebpf/2019/07/02/bpf-verifier-complexity.html and found eBPF [docs](https://ebpf.io/what-is-ebpf/) saying _"Program must have a finite complexity. The verifier will evaluate all possible execution paths and must be capable of completing the analysis within the limits of the configured upper complexity limit."_

As it relates to cilium and certain features, I'm guessing that for the generated eBPF code the cyclomatic complexity (and program size) for certain features is larger than a preset buffer size or complexity threshold that the kernel will accept.  Is this close?

**Comment by brb** [comment]
If we merge https://github.com/cilium/cilium/pull/15422, then this is definitely a release blocker. Added the label accordingly.

**Comment by jrajahalme** [comment]
Similar issue in the Kind GH action testing encryption at https://github.com/cilium/cilium-cli/pull/158:
```
2021-04-14T07:41:28.238252783Z level=info msg="  --kube-proxy-replacement='disabled'" subsys=daemon
2021-04-14T07:41:28.238514286Z level=info msg="Cilium 1.9.5 079bdaf 2021-03-10T13:12:19-08:00 go version go1.15.8 linux/amd64" subsys=daemon
2021-04-14T07:41:28.237561674Z level=info msg="  --encrypt-interface=''" subsys=daemon
2021-04-14T07:41:28.237564874Z level=info msg="  --encrypt-node='false'" subsys=daemon
2021-04-14T07:41:28.237723676Z level=info msg="  --ipam='kubernetes'" subsys=daemon
2021-04-14T07:41:28.276874052Z level=info msg="clang (10.0.0) and kernel (5.4.0) versions: OK!" subsys=linux-datapath
2021-04-14T07:46:29.491933327Z level=info msg="New encryption keys reclaiming SPI" spi=0 subsys=ipsec
...
2021-04-14T07:42:08.250629896Z level=info msg="Serving cilium health API at unix:///var/run/cilium/health.sock" subsys=health-server
2021-04-14T07:42:21.113832600Z level=error msg="Command execution failed" cmd="[tc filter replace dev lxc5782f00f5fb1 ingress prio 1 handle 1 bpf da obj 3081_next/bpf_lxc.o sec from-container]" error="exit status 1" subsys=datapath-loader
2021-04-14T07:42:21.114395307Z level=warning msg="Log buffer too small to dump verifier log 16777215 bytes (10 tries)!" subsys=datapath-loader
2021-04-14T07:42:21.114406107Z level=warning msg="JIT disabled, but 4/4 tail call maps in the program have JITed owner!" subsys=datapath-loader
2021-04-14T07:42:21.114410807Z level=warning msg="Error fetching program/map!" subsys=datapath-loader
2021-04-14T07:42:21.114415307Z level=warning msg="Unable to load program" subsys=datapath-loader
2021-04-14T07:42:21.126206251Z level=warning msg="JoinEP: Failed to load program" containerID=394314095a datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3081 error="Failed to load tc filter: exit status 1" file-path=3081_next/bpf_lxc.o identity=27403 ipv4=10.244.1.26 ipv6= k8sPodName=cilium-test/echo-same-node-66589c9569-vmr6p subsys=datapath-loader veth=lxc5782f00f5fb1
2021-04-14T07:42:21.126328253Z level=error msg="Error while rewriting endpoint BPF program" containerID=394314095a datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3081 error="Failed to load tc filter: exit status 1" identity=27403 ipv4=10.244.1.26 ipv6= k8sPodName=cilium-test/echo-same-node-66589c9569-vmr6p subsys=endpoint
2021-04-14T07:42:21.126452055Z level=warning msg="generating BPF for endpoint failed, keeping stale directory." containerID=394314095a datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3081 file-path=3081_next_fail identity=27403 ipv4=10.244.1.26 ipv6= k8sPodName=cilium-test/echo-same-node-66589c9569-vmr6p subsys=endpoint
2021-04-14T07:42:21.126640757Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=1.295833427s bpfLoadProg=44.693748346s bpfWaitForELF=1.296197031s bpfWriteELF="168.802µs" containerID=394314095a datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3081 error="Failed to load tc filter: exit status 1" identity=27403 ipv4=10.244.1.26 ipv6= k8sPodName=cilium-test/echo-same-node-66589c9569-vmr6p mapSync="33.5µs" policyCalculation="37.2µs" prepareBuild=2.061726ms proxyConfiguration="14.6µs" proxyPolicyCalculation="24.001µs" proxyWaitForAck=0s reason="syncing state to host" subsys=endpoint total=45.993888926s waitingForCTClean=400ns waitingForLock="13.4µs"
2021-04-14T07:42:21.126825959Z level=error msg="endpoint regeneration failed" containerID=394314095a datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3081 error="Failed to load tc filter: exit status 1" identity=27403 ipv4=10.244.1.26 ipv6= k8sPodName=cilium-test/echo-same-node-66589c9569-vmr6p subsys=endpoint
2021-04-14T07:42:21.126904060Z level=warning msg="Failed while regenerating endpoint" endpointID=3081 subsys=endpoint
```

**Comment by joestringer** [comment]
> "Log buffer too small to dump verifier log 16777215 bytes (10 tries)!"

In the past to investigate issues with this log message, we have manually edited the iproute2 to increase the maximum size of this buffer (basically grep for the log message in https://github.com/cilium/iproute2, edit the # of tries or base buffer size), included these changes into a new Cilium image and re-run the test to gather the actual verifier output. Might be one option to start to tackle this issue.

**Comment by pchaigno** [comment]
With @kkourt, we were able to reproduce on v1.9.3 and v1.9.6 (latest v1.9 stable release) on 5.4 kernels when `kubeProxyReplacement=disabled` and IPSec is enabled. The issue doesn't occur when IPSec is disabled. Note kernels 5.10+ are affected by a different complexity issue (https://github.com/cilium/cilium/issues/14726) that doesn't require IPSec to occur.

Master (7b27a99) and v1.10.0-rc1 are fine so removing the `release-blocker/1.10` label. Priority for this is still high as more folks will likely hit it on v1.9 as they upgrade their kernels.

Next step: one way to fix it would probably be to look at the differences between v1.10 and v1.9 (what did we not backport?) to figure out what helped.

**Comment by phuongnd96** [comment]
I'm facing problem with cilium v1.10.4, kubeProxyReplacement=strict + IPSec and kernel 4.18 too. Can not enable IPSec somehow

**Comment by pchaigno** [comment]
> I'm facing problem with cilium v1.10.4, kubeProxyReplacement=strict + IPSec and kernel 4.18 too.

@phuongnd96 IPsec is not supported with kubeProxyReplacement=strict. The Cilium agents should fatal with an error message if you try to enable both. Did you mean kubeProxyReplacement=disabled?

**Comment by pchaigno** [comment]
@msibrichins I think this bug may have been fixed in v1.9.12 and v1.10.6 by https://github.com/cilium/cilium/pull/17573. Could you give a try to either of those versions?

**Comment by pchaigno** [comment]
Closing as this should now be fixed.
